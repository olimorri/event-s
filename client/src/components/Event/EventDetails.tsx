import React from 'react';
import { useParams } from 'react-router-dom';
import Spinner from '../Handling/Spinner';
import { Flex, Stack, Text, Image, Box, Wrap } from '@chakra-ui/react';
import Attend from './Buttons/Attend';
import Unattend from './Buttons/Unattend';
import moment from 'moment';
import { motion } from 'framer-motion';
import {Event} from '../../interfaces/Event'

type EventDetailProps = {
  events: Event[],
  signUpDown: Function,
  user: any
}

export default function EventDetails({ events, signUpDown, user }: EventDetailProps) {
  let { id }: {id: string} = useParams();
  const event: Event | undefined = events.find((el: Event) => el._id === id); //Event: {}
  const handleSubmit = (e: MouseEvent) => {
    e.preventDefault();
    signUpDown('up', id);
  };
  const unattend = (e: MouseEvent) => {
    e.preventDefault();
    signUpDown('down', id);
  };
  console.log(user.eventList);
  return (
    <Flex
      minH={'100vh'}
      justify={'center'}
      bg={'custom.100'}
      bgImage="url('https://res.cloudinary.com/dujun1hoe/image/upload/v1615228154/event-s/gradient-background-26046-26731-hd-wallpapers.jpg_cenrqe.png')"
      bgSize="100% 50%"
      backgroundRepeat="no-repeat"
    >
      {!event ? (
        <Spinner />
      ) : (
        <Box
          w={'60%'}
          minW={'60%'}
          maxW={'60%'}
          bg={'gray.50'}
          marginTop={50}
          borderRadius="md"
          boxShadow={'md'}
          m={20}
          h={'60vw'}
          maxH={'80vw'}
        >
          <motion.div
            initial={{ opacity: 0 }}
            animate={{
              opacity: 1,
              transition: { delay: 0.4 },
            }}
          >
            <Wrap w={'100%'} align={'center'} p={1}>
              <Box w={'50%'} minW={'300px'}>
                <Image
                  borderRadius="md"
                  boxShadow="md"
                  src={event.photo}
                  alt=""
                />
              </Box>
              <Stack w={'50%'} align={'center'}>
                <Flex flexDirection={'column'} justify={'center'}>
                  <Text fontSize="3xl" m={3}>
                    {event.name}
                  </Text>
                  <Text align={'center'} fontSize="xl">
                    {event.location}
                  </Text>
                </Flex>
                {!Object.keys(user).length ||
                user.host ? null : !user.eventList.some(
                    (el: Event) => el._id === id,
                  ) ? (
                  <Attend handleSubmit={handleSubmit} />
                ) : (
                  <Unattend unattend={unattend} />
                )}
              </Stack>
            </Wrap>
            <Box p={4}>
              <Stack
                w={'100%'}
                h={'60vh'}
                direction={{ base: 'column', md: 'row' }}
              >
                <Flex
                  h={'60%'}
                  flex={1}
                  pl={4}
                  align={'end'}
                  justifyContent={'space-evenly'}
                  flexDirection={'column'}
                >
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.5,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/calendar-2_hmjf1m.png'
                        }
                        alt=""
                      />
                      <Text justify={'center'}>
                        {moment(event.date).format('MMM Do YY')}
                      </Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.6,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296946/event-s/icons/alarm-2_csdsdf.png'
                        }
                        alt=""
                      />
                      <Text>{event.duration}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.7,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615297028/event-s/icons/announcement_pr7pea.png'
                        }
                        alt=""
                      />
                      <Text>{event.limit}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.8,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/event_h6dtmv.png'
                        }
                        alt=""
                      />
                      <Text>{event.type}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.9,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/done_r4quc3.png'
                        }
                        alt=""
                      />
                      <Text>{event.attendees}</Text>
                    </Wrap>
                  </motion.div>
                </Flex>
                <Flex
                  h={'100%'}
                  flex={2}
                  p={4}
                  alignContent={'center'}
                  flexDirection={'column'}
                >
                  <Text fontWeight="bold" align={'center'}>
                    About this event
                  </Text>
                  <motion.div
                    initial={{
                      opacity: 0,
                      scale: 0,
                    }}
                    animate={{
                      opacity: 1,
                      scale: 1,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 1,
                      duration: 1,
                    }}
                  >
                    <Text mt={10} align={'center'}>
                      {event.description.slice(0, 1200)}...
                    </Text>
                  </motion.div>
                </Flex>

                <Stack spacing={4}></Stack>
              </Stack>
            </Box>
          </motion.div>
        </Box>
      )}
    </Flex>
  );
}
