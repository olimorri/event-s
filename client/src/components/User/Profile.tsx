/* eslint-disable no-unused-vars */
import React, { useEffect } from 'react';
import UsersApiService from '../../services/UsersApiService';
import { Link as RouterLink } from 'react-router-dom';
import { User } from '../../interfaces/User';

import {
  Flex,
  Box,
  Text,
  Button,
  Image,
  Heading,
  Wrap,
  Stack,
} from '@chakra-ui/react';
import { motion } from 'framer-motion';

export default function Profile(props: any) {
  const user: User = props.user;
  const setUser: Function = props.setUser;
  useEffect(() => {
    const getProfile = async () => {
      const userInfo = await UsersApiService.profile();

      function isUser(userInfo: any): userInfo is User {
        return userInfo.error === undefined;
      }

      if (isUser(userInfo)) {
        setUser(userInfo);
      } else {
        console.log('No user info found ðŸ˜ž');
      }
    };
    getProfile();
  }, [setUser]);

  return (
    <Box h={'100%'}>
      <Stack
        minH={'100vh'}
        bgImage="url('https://res.cloudinary.com/dujun1hoe/image/upload/v1615228154/event-s/gradient-background-26046-26731-hd-wallpapers.jpg_cenrqe.png')"
        bgSize="100% 50%"
        // bgSize="cover"
        backgroundRepeat="no-repeat"
        justify={'center'}
        align={'center'}
      >
        <motion.div
          initial={{
            opacity: 0,
            scale: 0,
          }}
          animate={{
            opacity: 1,
            scale: 1,
          }}
          transition={{
            // type: 'spring',
            // stiffness: 100,
            // @ts-ignore
            ease: 'easeIn',
            duration: 1,
          }}
        >
          <Flex
            borderRadius="md"
            bg={'custom.100'}
            flexDirection={'column'}
            maxW={'900px'}
            minH={'60%'}
            minW={'300px'}
            p={8}
          >
            {user.firstName ? (
              <>
                <Flex
                  w={'100%'}
                  justify={'center'}
                  p={4}
                  spacing="0px"
                  maxH={'300px'}
                >
                  <Box boxSize="sm" h={'100%'}>
                    <motion.div
                      initial={{
                        opacity: 0,
                        x: '-100vw',
                      }}
                      animate={{
                        opacity: 1,
                        x: 0,
                      }}
                      transition={{
                        //type: 'spring',
                        // stiffness: 100,
                        // @ts-ignore
                        ease: 'easeIn',
                        delay: 0.5,
                        duration: 1,
                      }}
                    >
                      <Image
                        alignSelf={'center'}
                        boxShadow={'md'}
                        borderRadius="md"
                        src={user.photo}
                        alt=""
                        minW={'300px'}
                      />
                    </motion.div>
                  </Box>
                  {/* <Spacer w={'30px'} /> */}
                  <Flex
                    w={'40vh'}
                    align={'center'}
                    justify={'center'}
                    flexDirection={'column'}
                  >
                    <Heading align={'center'}>Welcome </Heading>
                    <motion.div
                      initial={{
                        opacity: 0,
                        scale: 0,
                      }}
                      animate={{
                        opacity: 1,
                        scale: 1,
                      }}
                      transition={{
                        //type: 'spring',
                        stiffness: 100,
                        // @ts-ignore
                        ease: 'easeIn',
                        delay: 2,
                        duration: 4,
                      }}
                    >
                      <Heading align={'center'}>
                        {' '}
                        {user.firstName} {user.lastName}
                      </Heading>
                    </motion.div>
                  </Flex>
                </Flex>
                <Flex
                  m={5}
                  align={'center'}
                  justify={'center'}
                  dir={'row'}
                  justifyContent={'space-evenly'}
                >
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      //type: 'spring',
                      // stiffness: 100,
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.7,
                      duration: 1,
                    }}
                  >
                    <Box align={'center'}>
                      <Image
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615239488/event-s/account_ee8lc9.png'
                        }
                      />
                      <Text>User type:</Text>
                      <Text fontSize="2xl" fontWeight="bold">
                        {user.host ? 'Host' : 'Guest'}
                      </Text>
                    </Box>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      //type: 'spring',
                      // stiffness: 100,
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.9,
                      duration: 1,
                    }}
                  >
                    <Box align={'center'}>
                      <Image
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615239489/event-s/calendar_qsoxdg.png'
                        }
                      />
                      <Text>Number of events:</Text>
                      <Text fontSize="2xl" fontWeight="bold">
                        {user.eventList.length ? user.eventList.length : 0}
                      </Text>
                    </Box>
                  </motion.div>
                </Flex>

                {user.eventList.length ? (
                  <>
                    <Text align={'center'} justify={'center'}>
                      Current events:
                    </Text>
                    <Wrap
                      justify={'center'}
                      h={'100%'}
                      p={4}
                      marginTop={'5%'}
                      w={'100%'}
                      justifyContent={'space-around'}
                      borderRadius="md"
                    >
                      {user.eventList.map((el) => (
                        <motion.div
                          initial={{
                            opacity: 0,
                            scale: 0,
                          }}
                          animate={{
                            opacity: 1,
                            scale: 1,
                          }}
                          transition={{
                            //type: 'spring',
                            stiffness: 100,
                            // @ts-ignore
                            ease: 'easeIn',
                            delay: 1,
                            duration: 1,
                          }}
                        >
                          <RouterLink to={`/events/${el._id}`} key={el._id}>
                            <Box borderRadius="md" h={'100%'}>
                              <Button m={4}>{el.name}</Button>
                            </Box>
                          </RouterLink>
                        </motion.div>
                      ))}
                    </Wrap>
                  </>
                ) : (
                  <Text m={10} align={'center'} justify={'center'}>
                    You haven't got any events yet.{' '}
                    {user.host
                      ? `It's the perfect time to create one!`
                      : `Why don't you have a look around and sign up to some?`}
                  </Text>
                )}
              </>
            ) : (
              <RouterLink to="/login">
                <Button>You need to log in first</Button>
              </RouterLink>
            )}
          </Flex>
        </motion.div>
      </Stack>
    </Box>
  );
}
