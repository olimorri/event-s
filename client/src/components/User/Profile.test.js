import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import Profile from './Profile';
import EventDetails from '../Event/EventDetails';
import { BrowserRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

afterEach(cleanup);

const event1 = {
  attendees: 1,
  date: '2021-03-23T00:00:00.000Z',
  description: 'This is a test event',
  duration: 1,
  geolocation: '0.261773,51.116312',
  limit: 1,
  list: ['604a36891b95d669d99ebe2e'],
  location: 'tn25hs',
  name: 'test event 1',
  owner: '604b5c00ca29f24f751f492c',
  photo: 'https://source.unsplash.com/random/300x300',
  type: 'Food',
  _id: '604b5c00ca29f24f751f492c',
};

const event2 = {
  attendees: 2,
  date: '2021-03-24T00:00:00.000Z',
  description: 'This is also a test event',
  duration: 2,
  geolocation: '0.261773,51.116312',
  limit: 2,
  list: ['604a36891b95d669d99ebe2f'],
  location: 'tn25hd',
  name: 'test event 2',
  owner: '604b5c00ca29f24f751f492c',
  photo: 'https://source.unsplash.com/random/300x300',
  type: 'Other',
  _id: '604b5c00ca29f24f751f492c',
};

const user = {
  email: 'test@email.com',
  eventList: [event1, event2],
  firstName: 'Bob',
  lastName: 'Bobson',
  host: true,
  password: 'test',
  photo: 'https://unsplash.com/photos/mEBOHfzuPHk',
};

describe('Snapshot', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Unit Testing', () => {
  test('renders first name', () => {
    render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    const firstName = screen.getByText(/bob/i);
    expect(firstName).toBeInTheDocument();
  });

  test('renders last name', () => {
    render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    const lastName = screen.getByText(/bobson/i);
    expect(lastName).toBeInTheDocument();
  });

  test('renders user type', () => {
    render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    const userType = screen.getByText(/host/i);
    expect(userType).toBeInTheDocument();
  });

  test('renders event list', () => {
    render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    const event1 = screen.getByText(/^test event 1$/i);
    const event2 = screen.getByText(/^test event 2$/i);
    expect(event1).toBeInTheDocument();
    expect(event2).toBeInTheDocument();
  });

  test('renders number of events in list', () => {
    render(
      <BrowserRouter>
        <Profile user={user} />
      </BrowserRouter>,
    );
    const numberOfEvents = screen.getByText(/^2$/i);
    expect(numberOfEvents).toBeInTheDocument();
  });

  test('redirects to correct event details page on event click', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <Profile user={user} />
      </Router>,
    );
    const myEvent = screen.getByText(/^test event 1$/i);
    expect(history.location).toHaveProperty('pathname', '/');
    expect(history.entries).toHaveLength(1);
    fireEvent.click(myEvent);
    expect(history.location).toHaveProperty(
      'pathname',
      '/events/604b5c00ca29f24f751f492c',
    );
    expect(history.entries).toHaveLength(2);
    expect(history.entries[1]).toHaveProperty(
      'pathname',
      '/events/604b5c00ca29f24f751f492c',
    );
  });
});
