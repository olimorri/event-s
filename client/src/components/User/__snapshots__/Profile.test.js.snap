// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-10klw3m"
  >
    <div
      class="chakra-stack css-15e4fml"
    >
      <div
        style="opacity: 0; transform: scale(0) translateZ(0);"
      >
        <div
          class="css-jckc1i"
        >
          <div
            class="css-2499gv"
            spacing="0px"
          >
            <div
              class="css-13zcs50"
            >
              <div
                style="opacity: 0; transform: translateX(-100vw) translateZ(0);"
              >
                <img
                  alt=""
                  class="chakra-image__placeholder css-1c8jk1n"
                />
              </div>
            </div>
            <div
              class="css-1n2y2el"
            >
              <h2
                align="center"
                class="chakra-heading css-0"
              >
                Welcome 
              </h2>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <h2
                  align="center"
                  class="chakra-heading css-0"
                >
                   Bob Bobson
                </h2>
              </div>
            </div>
          </div>
          <div
            class="css-abirks"
            dir="row"
          >
            <div
              style="opacity: 0; transform: translateX(-100vw) translateZ(0);"
            >
              <div
                align="center"
                class="css-0"
              >
                <img
                  class="chakra-image__placeholder css-0"
                />
                <p
                  class="chakra-text css-0"
                >
                  User type:
                </p>
                <p
                  class="chakra-text css-1n9r2z"
                >
                  Host
                </p>
              </div>
            </div>
            <div
              style="opacity: 0; transform: translateX(-100vw) translateZ(0);"
            >
              <div
                align="center"
                class="css-0"
              >
                <img
                  class="chakra-image__placeholder css-0"
                />
                <p
                  class="chakra-text css-0"
                >
                  Number of events:
                </p>
                <p
                  class="chakra-text css-1n9r2z"
                >
                  8
                </p>
              </div>
            </div>
          </div>
          <p
            class="chakra-text css-xi606m"
            justify="center"
          >
            Current events:
          </p>
          <div
            class="chakra-wrap css-tiophl"
          >
            <ul
              class="chakra-wrap__list css-1vhqsv9"
            >
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604a268300da7a8522af8b23"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 1
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604ce151b61ff67613b5b6d9"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 2
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604ce171b61ff67613b5b6da"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 3
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604ce3d6b61ff67613b5b6db"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 4
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604f8bd4cdb45af7cd1ca265"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 5
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604f8c11cdb45af7cd1ca266"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 6
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604f8c49cdb45af7cd1ca267"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 7
                    </button>
                  </div>
                </a>
              </div>
              <div
                style="opacity: 0; transform: scale(0) translateZ(0);"
              >
                <a
                  href="/events/604fa49085cdd6032dfdf5e0"
                >
                  <div
                    class="css-1yf3uup"
                  >
                    <button
                      class="chakra-button css-19k6s7r"
                      type="button"
                    >
                      Test Event 8
                    </button>
                  </div>
                </a>
              </div>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
